### ISSUE TITLE
Incorrect subgradient of theano.tensor.max

### ISSUE THREAD

**Issue by TobyPDE** [Theano/Theano#4568]
TL;DR The calculated gradient of the max-operation is not always in the subdifferential.

**Idea**
Let **x** _= (x_1,...,x_N)_ be a real vector and consider the function 
_f(x_1,...,x_N) = max(x_1,...,x_N) = max(g_1(x_1,...,x_N),...,g_N(x_1,...,x_N))_ 
where _g_i(x_1,...,x_N) = x_i_. Basic math tells us that the subdifferential of _f_ at any fixed point **x** is the convex hull of the gradients of the active projection functions _g_i_ [1]. For _i in {1,...,N}_ the gradient of _g_i_ is the standard basis vector _**e**_i_. 

**Example**
Let _N = 2_, then the subdifferential of _f(x_1,x_2) = max(x_1, x_2)_ at a point **x** with _x_1 = x_2_ is the line segment _{a * (1, 0) + (1 - a) * (0, 1) : a in [0, 1]}_. 

**Issue**
Evaluating the gradient of _max(x_1, x_2)_ at a point **x** with _x_1 = x_2_ returns the gradient _(1, 1)_ which is not in the subdifferential. 

```python
import theano
import theano.tensor as T
import numpy as np

X = T.vector()
z = T.max(X)
Df = theano.function(
    inputs=[X],
    outputs=T.grad(z, X)
)

x = np.array([1, 1]).astype(dtype=theano.config.floatX)
gradient = Df(x)
print(gradient) # -> [1. 1.]
```

**Proposed solution**
[Here](https://github.com/Theano/Theano/blob/baf5227a93b7ff1d8b8aa59010591f86ecab0d43/theano/tensor/signal/pool.py#L798), we need to take some convex combination of the corresponding entries of `gz`. A natural choice would be to take the center of the subdifferential, i.e. divide `gz` by the number of active arguments. 

_I am not entirely sure if my reasoning is correct. So, please let's discuss._

[1] See 3.4 https://see.stanford.edu/materials/lsocoee364b/01-subgradients_notes.pdf

**Comment by lamblin** [comment]
Since max is convex, I think what you say about the subgradient seems true, and even in general, it may be a good way of handling such ties.
However, for us, it did not seem to be worth the effort, potential slowdowns, and potential suprising behaviours in other cases. The gradient wrt each of the component is still consistent now with the numerical one, at least in one direction, and ties are rare, so it did not seem to be an issue in practice.

**Comment by TobyPDE** [comment]
I absolutely agree that this is a border case that should only occur exceptionally rarely in practice and I agree that the computational overhead of dividing by the number of active arguments might not be worth it. However, upon investigating whether or not a fix might be significant, I noticed that this case is treated differently depending on whether or not the `ìgnore_border` parameter is set to `True`. 

```python
import theano
import theano.tensor as T
import numpy as np
from theano.tensor.signal.pool import pool_2d

X = T.ftensor4()
f1 = pool_2d(X, (2, 2), ignore_border=True)
f2 = pool_2d(X, (2, 2), ignore_border=False)

Df1 = theano.function(
    inputs=[X],
    outputs=T.grad(T.sum(f1), X)
)
Df2 = theano.function(
    inputs=[X],
    outputs=T.grad(T.sum(f2), X)
)

x = np.array([1, 1, 1, 1]).reshape((1, 1, 2, 2)).astype(dtype=theano.config.floatX)

print(Df1(x))
#[[[[ 1.  0.]
#   [ 0.  0.]]]]

print(Df2(x))
#[[[[ 1.  1.]
#   [ 1.  1.]]]]
```

This potentially "can" make a difference. 

```python
import theano
import theano.tensor as T
import numpy as np
from theano.tensor.signal.pool import pool_2d

def optimize(x, grad_desc):
    steps = 0
    eps = 1e-8

    while True:
        steps += 1
        loss, x = grad_desc(x)
        if loss < eps:
            print("Converged (objective=%e, eps=%e) in %d steps" % (loss, eps, steps))
            break

X = T.ftensor4()
f1 = pool_2d(X, (2, 2), ignore_border=True)
f2 = pool_2d(X, (2, 2), ignore_border=False)

loss1 = T.sum(f1)
loss2 = T.sum(f2)

x0 = np.array([1, 1, 1, 1]).reshape((1, 1, 2, 2)).astype(dtype=theano.config.floatX)

grad_desc1 = theano.function(
    inputs=[X],
    outputs=[loss1, X - 0.1 * T.grad(loss1, X)]
)
grad_desc2 = theano.function(
    inputs=[X],
    outputs=[loss2, X - 0.1 * T.grad(loss2, X)]
)

optimize(x0, grad_desc1)
# Converged (objective=-7.450581e-08, eps=1.000000e-08) in 41 steps
optimize(x0, grad_desc2)
# Converged (objective=-7.450581e-08, eps=1.000000e-08) in 11 steps
```

Obviously, these are pathological cases that don't deserve a lot of attention. Therefore, I don't think any action should be taken and the ticket can be closed. I just pointed this out because I did not expect the change in behavior when I changed the corresponding parameter. 

_I initially started investigating the case because the SegNet architecture for semantic segmentation uses the pooling indices in order to perform sparse up-sampling of feature maps. One possible way of retrieving the indices is the gradient operation._

**Comment by lamblin** [comment]
Indeed, we should be consistent across devices, and regardless of `ignore_border`.
We should check how cuDNN handles that case as well.

It would seem easy to simply record how many times the maximum was reached, and just divide by that number, but it is harder when you have overlapping windows.
I would keep that ticket open, to document the issue, but with low priority for the moment.

**Comment by nomader** [comment]
I would just like to point out that the observations of @TobyPDE are only true when cuDNN is enabled. 
Without cuDNN, `ignore_border` does not seem to affect the gradient output, and the gradients computation is (relatively) correct: all input locations receive a gradient of 1. 

```
import theano
theano.config.dnn.enabled = 'False'
import theano.tensor as T
import numpy as np
from theano.tensor.signal.pool import pool_2d

X = T.ftensor4()
f1 = pool_2d(X, (2, 2), ignore_border=True)
f2 = pool_2d(X, (2, 2), ignore_border=False)

Df1 = theano.function(
    inputs=[X],
    outputs=T.grad(T.sum(f1), X)
)
Df2 = theano.function(
    inputs=[X],
    outputs=T.grad(T.sum(f2), X)
)

x = np.array([1, 1, 1, 1]).reshape((1, 1, 2, 2)).astype(dtype=theano.config.floatX)

print(Df1(x))
#[[[[ 1.  1.]
#   [ 1.  1.]]]]

print(Df2(x))
#[[[[ 1.  1.]
#   [ 1.  1.]]]]
```

Therefore, the "bug" seems to lie withing cuDNN. 

In my opinion, with the advent of new non-typical architectures like what-where autoencoder, unpooling layers, etc., it becomes more vital to compute the gradients of max pooling correctly: when there is a tie, all input locations should get a gradient = 1/number_of_elements_in_pool_kernel (i.e., in the above examples, the output should look like:
```
#[[[[ .25  .25]
#   [ .25  .25]]]]
```
Would be nice if a solution is on the horizon soon.
